### PR TITLE
PIL-802  - Adding contact info + Print on final CYA page

### PR DESCRIPTION
**Pull request by javed-iqbal1**

### CODE REVIEW

File: app/controllers/rfm/ContactDetailsCheckYourAnswersController.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+package controllers.rfm
+
+import com.google.inject.Inject
+import config.FrontendAppConfig
+import controllers.actions.{DataRequiredAction, DataRetrievalAction, RfmIdentifierAction}
+import models.Mode
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, MessagesControllerComponents}
+import uk.gov.hmrc.play.bootstrap.frontend.controller.FrontendBaseController
+import utils.RowStatus
```
Comment (AKKUM) [inline app/controllers/rfm/ContactDetailsCheckYourAnswersController.scala:26]: I can still see unused import not been removed. Please have look all the comments.

File: app/controllers/rfm/RfmContactCheckYourAnswersController.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        Ok(view(rfmCorporatePositionSummaryList, rfmPrimaryContactList, rfmSecondaryContactList, address))
+      } else {
+        Redirect(controllers.routes.JourneyRecoveryController.onPageLoad())
+      }
+    } else {
+      Redirect(controllers.routes.UnderConstructionController.onPageLoad)
+    }
+  }
+
+  def onSubmit(mode: Mode): Action[AnyContent] = rfmIdentify.async { implicit request =>
```
Comment (jassalrichy) [inline app/controllers/rfm/RfmContactCheckYourAnswersController.scala:79]: Can remove mode param.

File: test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+import play.api.inject.bind
+import play.api.libs.json.Json
+import play.api.test.FakeRequest
+import play.api.test.Helpers._
+import viewmodels.govuk.SummaryListFluency
+import models.rfm.CorporatePosition
+
+import java.time.LocalDate
+import scala.concurrent.Future
+
```
Comment (jassalrichy) [inline test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala:37]: Remove unused imports.

File: test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+class RfmContactCheckYourAnswersControllerSpec extends SpecBase with SummaryListFluency {
+
+  private val date = LocalDate.now()
+
+  private val nonUkAddress = NonUKAddress(
+    addressLine1 = "1 drive",
+    addressLine2 = None,
+    addressLine3 = "la la land",
```
Comment (jassalrichy) [inline test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala:45]: As code will be in 'The Open' can we replace 'la la land' with 'line3'

File: test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    .setOrException(RfmSecondaryEmailPage, "email@address.com")
+    .setOrException(RfmSecondaryPhonePreferencePage, true)
+    .setOrException(RfmSecondaryCapturePhonePage, "1234567891")
+    .setOrException(RfmContactAddressPage, NonUKAddress("line1", None, "line3", None, None, countryCode = "US"))
+
+  "Check Your Answers Controller" when {
+
+    "must redirect to correct view when rfm feature false" in {
+      val ua = emptyUserAnswers
+        .setOrException(RfmPrimaryContactNamePage, "sad")
```
Comment (jassalrichy) [inline test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala:111]: As code will be in 'The Open' can we replace 'sad' with 'first last'

File: test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+    "return OK and the correct view if an answer is provided to every New RFM ID journey questions" in {
+
+      val application = applicationBuilder(userAnswers = Some(rfmID))
+        .build()
+      running(application) {
+        val request = FakeRequest(GET, controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad.url)
+        val result  = route(application, request).value
+        status(result) mustEqual OK
+
+        contentAsString(result) must include("Filing member details")
```
Comment (jassalrichy) [inline test/controllers/rfm/RfmContactCheckYourAnswersControllerSpec.scala:150]: No need to check the fields values rendered, just need to check the label values.

File: conf/messages.en [hunk 7]
```diff
@@ -929,8 +930,9 @@ rfm.corporatePosition.upe = Ultimate parent entity (UPE)
 rfm.corporatePosition.newNfm = New nominated filing member
 rfm.corporatePosition.insetText = To replace the existing filing member for this account you must have access to the information required to report Pillar 2 top-up taxes on the group’s behalf.
 rfm.corporatePosition.error.required = Select if you are the ultimate parent entity or a new nominated filing member
+rfm.corporatePosition.change.hidden = what is your position is in the corporate structure of the group
 
-rfm.contact.heading.caption = Contact details
+rfm.contact.heading.caption = Contact detNails
```
Comment (jassalrichy) [inline conf/messages.en:935]: correct typo.

File: app/views/rfm/RfmContactCheckYourAnswersView.scala.html [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+@layout(pageTitle = titleNoForm(messages("rfm.contactCheckYourAnswers.title"))) {
+
+    @formHelper(action = controllers.rfm.routes.RfmContactCheckYourAnswersController.onSubmit) {
+
+        @heading(messages("rfm.contactCheckYourAnswers.heading"), classes = "govuk-heading-l")
+
+        @h2(messages("rfm.contactCheckYourAnswers.filingmember"), size = "m")
+        @govukSummaryList(listCorporatePosition)
+
+        @heading(messages("contactCheckYourAnswers.first"), "govuk-heading-m", "h2")
```
Comment (jassalrichy) [inline app/views/rfm/RfmContactCheckYourAnswersView.scala.html:43]: should we be using rfm.contactCheckYourAnswers.first etc

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 9]
```diff
@@ -119,4 +166,44 @@ class ReplaceFilingMemberNavigator @Inject() {
...
+    userAnswers
+      .get(RfmUkBasedPage)
+      .map { rfmUkBased =>
+        if (rfmUkBased) {
+          controllers.rfm.routes.RfmEntityTypeController.onPageLoad(NormalMode)
+        } else {
+          controllers.rfm.routes.RfmNameRegistrationController.onPageLoad(NormalMode)
+        }
+      }
+      .getOrElse(routes.JourneyRecoveryController.onPageLoad())
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:179]: Need a test for when redirected to Journey Recovery page.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 10]
```diff
@@ -32,37 +33,52 @@ class ReplaceFilingMemberNavigator @Inject() {
...
-    case RfmPillar2ReferencePage  => _ => securityQuestionsCheckYourAnswers
-    case RfmRegistrationDatePage  => _ => securityQuestionsCheckYourAnswers
-    case RfmNameRegistrationPage  => whichCheckYourAnswerPageRfmQuestions
-    case RfmRegisteredAddressPage => whichCheckYourAnswerPageRfmQuestions
-    case _                        => _ => controllers.rfm.routes.StartPageController.onPageLoad
+    case RfmPillar2ReferencePage         => _ => securityQuestionsCheckYourAnswers
+    case RfmRegistrationDatePage         => _ => securityQuestionsCheckYourAnswers
+    case RfmNameRegistrationPage         => whichCheckYourAnswerPageRfmQuestions
+    case RfmRegisteredAddressPage        => whichCheckYourAnswerPageRfmQuestions
+    case RfmCorporatePositionPage        => rfmCorporatePositionCheckRouteLogic
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:68]: Need to add a test for RfmCorporatePositionPage.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 11]
```diff
@@ -119,4 +166,44 @@ class ReplaceFilingMemberNavigator @Inject() {
...
+  private def rfmCorporatePosition(userAnswers: UserAnswers): Call =
+    userAnswers
+      .get(RfmCorporatePositionPage)
+      .map { corporatePosition =>
+        if (corporatePosition == CorporatePosition.Upe) {
+          controllers.rfm.routes.RfmContactDetailsRegistrationController.onPageLoad
+        } else if (corporatePosition == CorporatePosition.NewNfm) {
+          controllers.rfm.routes.CheckNewFilingMemberController.onPageLoad(NormalMode)
+        } else {
+          routes.JourneyRecoveryController.onPageLoad()
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:190]: Need a test for when redirected to Journey Recovery page.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 12]
```diff
@@ -119,4 +166,44 @@ class ReplaceFilingMemberNavigator @Inject() {
...
+        }
+      }
+      .getOrElse(routes.JourneyRecoveryController.onPageLoad())
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:193]: Need a test for when redirected to Journey Recovery page.

File: app/pages/RfmCorporatePositionPage.scala [hunk 13]
```diff
@@ -17,11 +17,32 @@
 package pages
 
 import models.rfm.CorporatePosition
+import models.UserAnswers
 import play.api.libs.json.JsPath
+import scala.util.Try
 
 case object RfmCorporatePositionPage extends QuestionPage[CorporatePosition] {
```
Comment (jassalrichy) [inline app/pages/RfmCorporatePositionPage.scala:24]: Need a page spec.

File: app/pages/RfmUkBasedPage.scala [hunk 14]
```diff
@@ -16,12 +16,33 @@
 
 package pages
 
+import models.UserAnswers
 import play.api.libs.json.JsPath
 
+import scala.util.Try
+
 case object RfmUkBasedPage extends QuestionPage[Boolean] {
```
Comment (jassalrichy) [inline app/pages/RfmUkBasedPage.scala:24]: Need a page spec.

File: app/viewmodels/checkAnswers/EntityTypePartnershipCompanyRegRfmSummary.scala [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object EntityTypePartnershipCompanyRegRfmSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers
+      .get(RfmGRSResponsePage)
+      .flatMap { GRS =>
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/EntityTypePartnershipCompanyRegRfmSummary.scala:32]: use lowercase grs

File: app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyUtrRfmSummary.scala [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object EntityTypeIncorporatedCompanyUtrRfmSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers
+      .get(RfmGRSResponsePage)
+      .flatMap { GRS =>
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyUtrRfmSummary.scala:33]: use lowercase grs.

File: app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyNameRfmSummary.scala [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object EntityTypeIncorporatedCompanyNameRfmSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers
+      .get(RfmGRSResponsePage)
+      .flatMap { GRS =>
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyNameRfmSummary.scala:33]: use lowercase grs.

File: app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyRegRfmSummary.scala [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+import uk.gov.hmrc.govukfrontend.views.viewmodels.summarylist.SummaryListRow
+import viewmodels.govuk.summarylist._
+import viewmodels.implicits._
+
+object EntityTypeIncorporatedCompanyRegRfmSummary {
+
+  def row(answers: UserAnswers)(implicit messages: Messages): Option[SummaryListRow] =
+    answers
+      .get(RfmGRSResponsePage)
+      .flatMap { GRS =>
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/EntityTypeIncorporatedCompanyRegRfmSummary.scala:33]: use lowercase grs.

File: app/viewmodels/checkAnswers/RfmContactAddressSummary.scala [hunk 19]
```diff
@@ -29,15 +29,15 @@ object RfmContactAddressSummary {
 
   def row(answers: UserAnswers, countryOptions: CountryOptions)(implicit messages: Messages): Option[SummaryListRow] =
     answers
-      .get(FmRegisteredAddressPage)
+      .get(RfmContactAddressPage)
       .map { answer =>
         val country = countryOptions.getCountryNameFromCode(answer.countryCode)
         SummaryListRowViewModel(
           key = "rfmContactAddress.checkYourAnswersLabel",
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/RfmContactAddressSummary.scala:36]: can we the message key names consistent by using prefix 'frm.'

File: app/viewmodels/checkAnswers/RfmSecondaryTelephoneSummary.scala [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package viewmodels.checkAnswers
+
+import models.{CheckMode, UserAnswers}
+import pages.{RfmSecondaryCapturePhonePage, SubSecondaryCapturePhonePage}
```
Comment (jassalrichy) [inline app/viewmodels/checkAnswers/RfmSecondaryTelephoneSummary.scala:20]: remove unused import.

File: test/controllers/rfm/RfmContactAddressControllerSpec.scala [hunk 21]
```diff
@@ -133,7 +133,7 @@ class RfmContactAddressControllerSpec extends SpecBase {
```
Comment (jassalrichy) [inline test/controllers/rfm/RfmContactAddressControllerSpec.scala:133]: Remove unused imports.

File: test/helpers/UserAnswersFixture.scala [hunk 22]
```diff
@@ -70,6 +71,62 @@ trait UserAnswersFixture extends TryValues {
     )
   )
 
+  val grsResponseRfm = GrsResponse(
```
Comment (jassalrichy) [inline test/helpers/UserAnswersFixture.scala:74]: Do we need to create a 'grsResponseRfm' ? Use the existing 'grsResponse'

File: app/viewmodels/checkAnswers/RfmContactAddressSummary.scala (hunk 19 above)
Comment (javed-iqbal1) [reply app/viewmodels/checkAnswers/RfmContactAddressSummary.scala:36]: it is starting with rfm, i think fine.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 23]
```diff
@@ -32,81 +33,116 @@ class ReplaceFilingMemberNavigator @Inject() {
...
+    case RfmUkBasedPage                  => rfmUkBasedCheckRouteLogic
+    case RfmPrimaryContactNamePage       => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+    case RfmPrimaryContactEmailPage      => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+    case RfmContactByTelephonePage       => telephonePreferenceLogic
+    case RfmCapturePrimaryTelephonePage  => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+    case RfmAddSecondaryContactPage      => rfmAddSecondaryContactRoute
+    case RfmSecondaryContactNamePage     => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+    case RfmSecondaryEmailPage           => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+    case RfmSecondaryPhonePreferencePage => rfmSecondaryPhonePreference
+    case RfmSecondaryCapturePhonePage    => _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:77]: Need a unit test for RfmSecondaryCapturePhonePage -> RfmContactCheckYourAnswersController

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 24]
```diff
@@ -32,81 +33,116 @@ class ReplaceFilingMemberNavigator @Inject() {
...
-        if (provided) {
+      .map {
+        case true if userAnswers.get(RfmCapturePrimaryTelephonePage).isEmpty =>
           controllers.rfm.routes.RfmCapturePrimaryTelephoneController.onPageLoad(NormalMode)
-        } else {
+        case true => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
+        case false if userAnswers.get(RfmAddSecondaryContactPage).isEmpty =>
           controllers.rfm.routes.RfmAddSecondaryContactController.onPageLoad(NormalMode)
-        }
+        case false => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:104]: Need a unit test for redirect when false.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 25]
```diff
@@ -32,81 +33,116 @@ class ReplaceFilingMemberNavigator @Inject() {
...
-        }
+        case _ => controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
       }
       .getOrElse(routes.JourneyRecoveryController.onPageLoad())
 
+  private def rfmSecondaryPhoneCaptureRoute(userAnswers: UserAnswers): Call =
+    if (userAnswers.get(RfmContactAddressPage).isEmpty) {
+      controllers.rfm.routes.RfmContactAddressController.onPageLoad(NormalMode)
+    } else {
+      controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:136]: Need a test for the else condition.

File: app/navigation/ReplaceFilingMemberNavigator.scala [hunk 26]
```diff
@@ -32,81 +33,116 @@ class ReplaceFilingMemberNavigator @Inject() {
...
+    }
+
+  private def rfmPrimaryPhoneCaptureRoute(userAnswers: UserAnswers): Call =
+    if (userAnswers.get(RfmAddSecondaryContactPage).isEmpty) {
+      controllers.rfm.routes.RfmAddSecondaryContactController.onPageLoad(NormalMode)
+    } else {
+      controllers.rfm.routes.RfmContactCheckYourAnswersController.onPageLoad
```
Comment (jassalrichy) [inline app/navigation/ReplaceFilingMemberNavigator.scala:143]: Need a test for the else condition.

File: test/helpers/UserAnswersFixture.scala (hunk 22 above)
Comment (aashik-varahan) [reply test/helpers/UserAnswersFixture.scala]: Removed redundant code grsResponseRfm

**Review by jassalrichy — APPROVED** [review]

File: test/navigation/ReplaceFilingMemberNavigatorSpec.scala
Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: Please check Normal mode routing for "go to contact address if secondary contact is not available" description looks incorrect. line 83.

Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: Correct "go to Add Secondary Contact details page where they enter their phone details" at line 111. to
"go to Add Secondary Contact details page **when** they enter their phone details"

Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: Please update "go to RfmSecondartyTelephonePreference page from RfmSecondaryTelephonePreference page" to
"go to RfmSecondartyTelephonePreference page from **RfmSecondaryEmailPage** page" on line 164

Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: Please update "go to RfmSecondaryTelephonePreference page if they choose to nominate a secondary contact number" to "go to **RfmSecondaryCapturePhonePage** page if they choose to nominate a secondary contact number" on line 169

Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: Please update "go to submit and review CYA page from secondary contact telephone page with address" to
"**go to RfmContactAddress page from secondary contact telephone page**"

Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:1]: This needs updating on line 327:
"go to **rfm UPE page** from corporate position if ultimate parent entity is selected"
line 333 should be:
submitAndReview

File: test/navigation/ReplaceFilingMemberNavigatorSpec.scala [hunk 27]
```diff
@@ -361,6 +367,23 @@ class ReplaceFilingMemberNavigatorSpec extends SpecBase {
           submitAndReview
       }
 
+      "go to submit and review CYA page from  do you want to add secondary contact  page with answer Yes" in {
```
Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:375]: "go to s**ubmit and review CYA** page from do you want to add secondary contact page with answer Yes" is incorrect as it is going to rfm secomndary contact name page.

File: test/navigation/ReplaceFilingMemberNavigatorSpec.scala [hunk 28]
```diff
@@ -361,6 +367,23 @@ class ReplaceFilingMemberNavigatorSpec extends SpecBase {
...
+        val ua = emptyUserAnswers
+          .setOrException(RfmAddSecondaryContactPage, true)
```
Comment (jassalrichy) [inline test/navigation/ReplaceFilingMemberNavigatorSpec.scala:377]: We only need to set RfmAddSecondaryContactPage to true.
No need to set values for RfmSecondaryContactNamePage, RfmSecondaryEmailPage, RfmSecondaryPhonePreferencePage.

**Review by jassalrichy — APPROVED** [review]

**Pull request merged**